packages/medusa/src/loaders/plugins.ts

```diff
@@ -118,47 +142,89 @@ async function runLoaders(pluginDetails: PluginDetails, container: MedusaContain
   )
 }
 
-function registerMedusaApi(pluginDetails: PluginDetails, container: MedusaContainer): void {
+async function registerMedusaApi(
+  pluginDetails: PluginDetails,
+  container: MedusaContainer
+): Promise<void> {
   registerMedusaMiddleware(pluginDetails, container)
-  registerStrategies(pluginDetails, container)
+  await registerStrategies(pluginDetails, container)
 }
 
-function registerStrategies(pluginDetails: PluginDetails, container: MedusaContainer): void {
-  let module
-  try {
-    const path = `${pluginDetails.resolve}/strategies/tax-calculation`
-    if (existsSync(path)) {
-      module = require(path).default
+async function registerStrategies(
+  pluginDetails: PluginDetails,
+  container: MedusaContainer
+): Promise<void> {
+  const files = glob.sync(`${pluginDetails.resolve}/strategies/[!__]*`, {})
+  const registeredServices = {}
+
+  const batchJobsStrategyFiles: string[] = []
+
+  files.map((file) => {
+    const module = require(file).default
+
+    if (isTaxCalculationStrategy(module.prototype)) {
+      if (!("taxCalculationStrategy" in registeredServices)) {
+        container.register({
+          taxCalculationStrategy: asFunction(
+            (cradle) => new module(cradle, pluginDetails.options)
+          ).singleton(),
+        })
+        registeredServices["taxCalculationStrategy"] = file
+      } else {
+        logger.warn(
+          `Cannot register ${file}. A tax calculation strategy is already registered`
+        )
+      }
+    } else if (isBatchJobStrategy(module.prototype)) {
+      batchJobsStrategyFiles.push(file)
+    } else if (isPriceSelectionStrategy(module.prototype)) {
+      if (!("priceSelectionStrategy" in registeredServices)) {
+        container.register({
+          priceSelectionStrategy: asFunction(
+            (cradle) => new module(cradle, pluginDetails.options)
+          ).singleton(),
+        })
+
+        registeredServices["priceSelectionStrategy"] = file
+      } else {
+        logger.warn(
+          `Cannot register ${file}. A price selection strategy is already registered`
+        )
+      }
     } else {
-      return
+      const logger = container.resolve<Logger>("logger")
+      logger.warn(
+        `${file} did not export a class that implements a strategy interface. Your Medusa server will still work, but if you have written custom strategy logic it will not be used. Make sure to implement the proper interface.`
+      )
     }
-  } catch (err) {
-    return
-  }
+  })
 
-  if (isTaxCalculationStrategy(module.prototype)) {
+  if (batchJobsStrategyFiles.length > 0) {
     container.register({
-      taxCalculationStrategy: asFunction(
-        (cradle) => new module(cradle, pluginDetails.options)
+      batchJobStrategies: asFunction((cradle) =>
+        batchJobsStrategyFiles.map((bs) => {
+          const module = require(bs).default
+
+          return new module(cradle, pluginDetails.options)
+        })
```

Comment:
todo: use the `registerAdd` method on the cradle to dynamically fill the array. 

It might also make sense to split the strategies across `batchType`. This way when we need to find the correct BatchJobStrategy for a given batch job of type x we would be able to check with the container on the field: `batch_${job.type}`

wdyt?

Reply:
I think splitting handlers by type in some way makes the most sense in order to make processing easier for each type of batch (especially when we end up having many batch jobs).

Reply:
@srindom 

I have done the following: 
- All batch job strategies are registered in an array under the `batchJobStrategies` key
- All batch job strategies with the same batch-type is registered under the name of the batch type (in case the batch type contains a hyphen it can be retrieved as such: `constructor({ "batch-type": batchType })`
- All batch job strategies are registered individually by their name and identifier (with the prefix `bs_`) as done with other providers